File: ts/tests/assumptions.test.ts
/**
 * Assumption violation conformance tests.
 *
 * These tests verify that assumption violations are reported correctly and
 * consistently across all languages. The test data is loaded from shared
 * JSON files in tests/assumptions/.
 */

import * as fs from 'fs';
import * as path from 'path';
import {
  center,
  ratio,
  relSpread,
  spread,
  shift,
  avgSpread,
  disparity,
  centerBounds,
} from '../src/estimators';
import { signedRankMargin } from '../src/signedRankMargin';
import { AssumptionError } from '../src/assumptions';

interface ExpectedViolation {
  id: string;
}

interface TestInputs {
  x?: (number | string)[];
  y?: (number | string)[];
  misrate?: number | string;
  n?: number;
  seed?: string;
}

interface AssumptionTestCase {
  name: string;
  function: string;
  inputs: TestInputs;
  expected_violation: ExpectedViolation;
}

interface AssumptionTestSuite {
  suite: string;
  description: string;
  cases: AssumptionTestCase[];
}

interface SuiteEntry {
  name: string;
  file: string;
  description: string;
}

interface Manifest {
  name: string;
  description: string;
  suites: SuiteEntry[];
}

/**
 * Parses a JSON value into a number, handling special values.
 */
function parseValue(v: number | string): number {
  if (typeof v === 'number') {
    return v;
  }
  if (v === 'NaN') {
    return NaN;
  }
  if (v === 'Infinity') {
    return Infinity;
  }
  if (v === '-Infinity') {
    return -Infinity;
  }
  throw new Error(`Unknown string value: ${v}`);
}

/**
 * Parses an array of JSON values into a number array.
 */
function parseArray(arr: (number | string)[] | undefined): number[] {
  if (arr === undefined) {
    return [];
  }
  return arr.map(parseValue);
}

/**
 * Dispatches to the appropriate estimator function.
 */
function callFunction(funcName: string, inputs: TestInputs): void {
  const x = parseArray(inputs.x);
  const y = parseArray(inputs.y);

  switch (funcName) {
    case 'Center':
      center(x);
      break;
    case 'Ratio':
      ratio(x, y);
      break;
    case 'RelSpread':
      relSpread(x);
      break;
    case 'Spread':
      spread(x);
      break;
    case 'Shift':
      shift(x, y);
      break;
    case 'AvgSpread':
      avgSpread(x, y);
      break;
    case 'Disparity':
      disparity(x, y);
      break;
    case 'CenterBounds':
      centerBounds(x, parseValue(inputs.misrate!));
      break;
    case 'SignedRankMargin':
      signedRankMargin(inputs.n!, parseValue(inputs.misrate!));
      break;
    default:
      throw new Error(`Unknown function: ${funcName}`);
  }
}

/**
 * Loads all assumption test cases from the shared test data.
 */
function loadAssumptionTestCases(): { suiteName: string; testCase: AssumptionTestCase }[] {
  const repoRoot = path.join(__dirname, '../..');
  const assumptionsDir = path.join(repoRoot, 'tests', 'assumptions');

  const manifestPath = path.join(assumptionsDir, 'manifest.json');
  const manifest: Manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf-8'));

  const testCases: { suiteName: string; testCase: AssumptionTestCase }[] = [];

  for (const suiteEntry of manifest.suites) {
    const suitePath = path.join(assumptionsDir, suiteEntry.file);
    const suite: AssumptionTestSuite = JSON.parse(fs.readFileSync(suitePath, 'utf-8'));

    for (const testCase of suite.cases) {
      testCases.push({ suiteName: suite.suite, testCase });
    }
  }

  return testCases;
}

describe('Assumption Violation Tests', () => {
  const testCases = loadAssumptionTestCases();

  test.each(testCases)('$suiteName/$testCase.name', ({ testCase }) => {
    const expectedId = testCase.expected_violation.id;

    let thrownError: AssumptionError | null = null;
    try {
      callFunction(testCase.function, testCase.inputs);
    } catch (e) {
      if (e instanceof AssumptionError) {
        thrownError = e;
      } else {
        throw e;
      }
    }
    expect(thrownError).not.toBeNull();
    expect(thrownError!.violation.id).toBe(expectedId);
  });
});

File: go/assumptions_test.go
package pragmastat

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"testing"
)

// AssumptionTestInputs represents input data for assumption tests.
type AssumptionTestInputs struct {
	X       []interface{} `json:"x,omitempty"`
	Y       []interface{} `json:"y,omitempty"`
	Misrate any           `json:"misrate,omitempty"`
	N       *int          `json:"n,omitempty"`
	Seed    *string       `json:"seed,omitempty"`
}

// ExpectedViolation represents the expected violation.
type ExpectedViolation struct {
	ID string `json:"id"`
}

// AssumptionTestCase represents a single test case.
type AssumptionTestCase struct {
	Name              string               `json:"name"`
	Function          string               `json:"function"`
	Inputs            AssumptionTestInputs `json:"inputs"`
	ExpectedViolation ExpectedViolation    `json:"expected_violation"`
}

// AssumptionTestSuite represents a test suite.
type AssumptionTestSuite struct {
	Suite       string               `json:"suite"`
	Description string               `json:"description"`
	Cases       []AssumptionTestCase `json:"cases"`
}

// SuiteEntry represents an entry in the manifest.
type SuiteEntry struct {
	Name        string `json:"name"`
	File        string `json:"file"`
	Description string `json:"description"`
}

// AssumptionManifest represents the manifest of test suites.
type AssumptionManifest struct {
	Name        string       `json:"name"`
	Description string       `json:"description"`
	Suites      []SuiteEntry `json:"suites"`
}

// parseValue converts a JSON value to float64, handling special values.
func parseValue(v interface{}) float64 {
	switch val := v.(type) {
	case float64:
		return val
	case int:
		return float64(val)
	case string:
		switch val {
		case "NaN":
			return math.NaN()
		case "Infinity":
			return math.Inf(1)
		case "-Infinity":
			return math.Inf(-1)
		default:
			panic("Unknown string value: " + val)
		}
	default:
		panic("Unexpected value type")
	}
}

// parseArray converts a slice of interface{} to []float64.
func parseArray(arr []interface{}) []float64 {
	if arr == nil {
		return []float64{}
	}
	result := make([]float64, len(arr))
	for i, v := range arr {
		result[i] = parseValue(v)
	}
	return result
}

// callAssumptionFunction dispatches to the appropriate estimator function.
func callAssumptionFunction(funcName string, inputs AssumptionTestInputs) error {
	x := parseArray(inputs.X)
	y := parseArray(inputs.Y)

	switch funcName {
	case "Center":
		_, err := Center(x)
		return err
	case "Ratio":
		_, err := Ratio(x, y)
		return err
	case "RelSpread":
		_, err := RelSpread(x)
		return err
	case "Spread":
		_, err := Spread(x)
		return err
	case "Shift":
		_, err := Shift(x, y)
		return err
	case "AvgSpread":
		_, err := AvgSpread(x, y)
		return err
	case "Disparity":
		_, err := Disparity(x, y)
		return err
	case "MedianBounds":
		_, err := MedianBounds(x, parseValue(inputs.Misrate))
		return err
	case "CenterBounds":
		_, err := CenterBounds(x, parseValue(inputs.Misrate))
		return err
	case "CenterBoundsApprox":
		_, err := CenterBoundsApproxWithSeed(x, parseValue(inputs.Misrate), inputs.Seed)
		return err
	case "SignedRankMargin":
		_, err := SignedRankMargin(*inputs.N, parseValue(inputs.Misrate))
		return err
	default:
		panic("Unknown function: " + funcName)
	}
}

func TestAssumptionViolations(t *testing.T) {
	assumptionsDir := filepath.Join("..", "tests", "assumptions")

	// Load manifest
	manifestPath := filepath.Join(assumptionsDir, "manifest.json")
	manifestData, err := os.ReadFile(manifestPath)
	if err != nil {
		t.Fatalf("Failed to read manifest: %v", err)
	}

	var manifest AssumptionManifest
	if err := json.Unmarshal(manifestData, &manifest); err != nil {
		t.Fatalf("Failed to parse manifest: %v", err)
	}

	// Run each suite
	for _, suiteEntry := range manifest.Suites {
		suitePath := filepath.Join(assumptionsDir, suiteEntry.File)
		suiteData, err := os.ReadFile(suitePath)
		if err != nil {
			t.Fatalf("Failed to read suite %s: %v", suiteEntry.Name, err)
		}

		var suite AssumptionTestSuite
		if err := json.Unmarshal(suiteData, &suite); err != nil {
			t.Fatalf("Failed to parse suite %s: %v", suiteEntry.Name, err)
		}

		for _, testCase := range suite.Cases {
			testName := suite.Suite + "/" + testCase.Name

			t.Run(testName, func(t *testing.T) {
				expectedID := AssumptionID(testCase.ExpectedViolation.ID)

				err := callAssumptionFunction(testCase.Function, testCase.Inputs)

				if err == nil {
					t.Errorf("Expected violation %s but got success", expectedID)
					return
				}

				assumptionErr, ok := err.(*AssumptionError)
				if !ok {
					t.Errorf("Expected AssumptionError but got %T: %v", err, err)
					return
				}

				if assumptionErr.Violation.ID != expectedID {
					t.Errorf("Expected id=%s, got %s",
						expectedID, assumptionErr.Violation.ID)
					return
				}
			})
		}
	}
}
